lib/index.ts

```diff
@@ -0,0 +1,471 @@
+	docker: import('dockerode');
+	// optional: cached result of dockerode's docker.version()
+	dockerVersionObj?: EngineVersion;
+}
+
+export class DockerProgress {
+	docker: import('dockerode');
+
+	protected reporter?: ProgressReporter;
+	protected engineVersion: EngineVersion | undefined;
+	private engineVersionPromise: Promise<EngineVersion> | undefined;
+
+	constructor(protected opts: DockerProgressOpts) {
+		this.docker = opts.docker;
+		this.engineVersion = opts.dockerVersionObj;
+	}
+
+	protected getProgressRenderer(
+		stepCount = DEFAULT_PROGRESS_BAR_STEP_COUNT,
+	): (percentage: number) => string {
+		return (percentage: number) => $renderProgress(percentage, stepCount);
+	}
+
+	protected async getProgressReporter(): Promise<ProgressReporter> {
+		if (this.reporter != null) {
+			return this.reporter;
+		}
+		const renderer = this.getProgressRenderer();
+		this.reporter = (await this.isBalenaEngine())
+			? new BalenaProgressReporter(renderer)
+			: new ProgressReporter(renderer);
+		return this.reporter;
+	}
+
+	aggregateProgress(
+		count: number,
+		onProgress: ProgressCallback,
+	): ProgressCallback[] {
+		const renderer = this.getProgressRenderer();
+		const states = _.times(count, () => ({
+			percentage: 0,
+		}));
+		return _.times(
+			count,
+			(index) =>
+				function (evt: any): void {
+					// update current reporter state
+					states[index].percentage = evt.percentage;
+					// update totals
+					const percentage = Math.floor(
+						_.sumBy(states, 'percentage') / (states.length || 1),
+					);
+					// update event
+					evt.totalProgress = renderer(percentage);
+					evt.percentage = percentage;
+					evt.progressIndex = index;
+					// call callback with aggregate event
+					onProgress(evt);
+				},
+		);
+	}
+
+	/** Pull docker image calling onProgress with extended progress info regularly */
+	async pull(
+		image: string,
+		onProgress: ProgressCallback,
+		options?: PullPushOptions,
+	): Promise<string> {
+		const ignoreErrorEvents = !!options?.ignoreProgressErrorEvents;
+		const reporter = await this.getProgressReporter();
+		onProgress = reporter.pullProgress(onProgress);
+		const stream = await this.docker.pull(image, options);
+		const hash = await awaitRegistryStream(
+			stream,
+			onProgress,
+			ignoreErrorEvents,
+		);
+		return hash;
+	}
+
+	/** Push docker image calling onProgress with extended progress info regularly */
+	async push(
+		image: string,
+		onProgress: ProgressCallback,
+		options?: PullPushOptions,
+	): Promise<string> {
+		const ignoreErrorEvents = !!options?.ignoreProgressErrorEvents;
+		const reporter = await this.getProgressReporter();
+		onProgress = reporter.pushProgress(onProgress);
+		const imageObj = this.docker.getImage(image);
+		const stream = await imageObj.push(options as ImagePushOptions);
+		const hash = await awaitRegistryStream(
+			stream,
+			onProgress,
+			ignoreErrorEvents,
+		);
+		return hash;
+	}
+
+	async isBalenaEngine(): Promise<boolean> {
```

Comment:
Removed `public` keyword to address https://github.com/balena-io-modules/docker-progress/pull/54#discussion_r626640433